[JIT] Put explicit error message on class attribute accesses.

zhxchen17 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#55723 [JIT] Put explicit error message on class attribute accesses.**

Summary: Resolving https://github.com/pytorch/pytorch/issues/51139

Test Plan: python test/test_jit.py TestClassType.test_unresolved_attributes

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27691960](https://our.internmc.facebook.com/intern/diff/D27691960)

zhxchen17 committed:
[JIT] Put explicit error message on class attribute accesses.

Summary: Resolving https://github.com/pytorch/pytorch/issues/51139

Test Plan: python test/test_jit.py TestClassType.test_unresolved_attributes

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

zhxchen17 committed:
Update on "[JIT] Put explicit error message on class attribute accesses."


Summary: Resolving https://github.com/pytorch/pytorch/issues/51139

Test Plan: python test/test_jit.py TestClassType.test_unresolved_attributes

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27691960](https://our.internmc.facebook.com/intern/diff/D27691960)

[ghstack-poisoned]

zhxchen17 committed:
Update on "[JIT] Put explicit error message on class attribute accesses."


Summary: Resolving https://github.com/pytorch/pytorch/issues/51139

Test Plan: python test/test_jit.py TestClassType.test_unresolved_attributes

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27691960](https://our.internmc.facebook.com/intern/diff/D27691960)

[ghstack-poisoned]

zhxchen17 committed:
Update on "[JIT] Put explicit error message on class attribute accesses."


Summary: Resolving https://github.com/pytorch/pytorch/issues/51139

Test Plan: python test/test_jit.py TestClassType.test_unresolved_attributes

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27691960](https://our.internmc.facebook.com/intern/diff/D27691960)

[ghstack-poisoned]

zhxchen17 committed:
Update on "[JIT] Put explicit error message on class attribute accesses."


Summary: Resolving https://github.com/pytorch/pytorch/issues/51139

Test Plan: python test/test_jit.py TestClassType.test_unresolved_attributes

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27691960](https://our.internmc.facebook.com/intern/diff/D27691960)

[ghstack-poisoned]

gmagogsfm commented on aten/src/ATen/core/type.cpp:
```diff
@@ -1174,6 +1174,13 @@ std::string ClassType::getForwardHookErrorMessage(int hook_idx) const {
   return return_string;
 }
 
+bool ClassType::hasUnresolvedClassAttribute(const std::string& name) const {
```
```suggestion
bool ClassType::isUnresolvedClassAttribute(const std::string& name) const {
```

zhxchen17 commented on test/jit/test_class_type.py:
```diff
@@ -1470,3 +1470,28 @@ def forward(self, x: int) -> int:
+                self.x = x
+                self._prop = 0
+
+            def getX(self):
```
50% is because I copy-paste another test :) , 50% is I want to ensure the mechanism works without crash under the presence of method and properties.

zhxchen17 commented on torch/jit/frontend.py:
```diff
@@ -156,6 +156,10 @@ def get_class_properties(cls, self_name):
 
     return properties
 
+def get_unresolved_class_attributes(cls, ast):
```
Currently this function returns something like `['__module__', 'attr', '__dict__', '__weakref__', '__doc__']` 
I think you mean we should get rid of builtin methods with __ prefix right?

To ensure all unresolved class attrs are what user specified in the class, I agree it's better to parse the class on our own instead of relying on python semantics.

Do we have a standalone parser for doing this? I'm not sure whether `inspect` module will return an ast.

zhxchen17 committed:
Update on "[JIT] Put explicit error message on class attribute accesses."


Summary: Resolving https://github.com/pytorch/pytorch/issues/51139

Test Plan: python test/test_jit.py TestClassType.test_unresolved_attributes

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27691960](https://our.internmc.facebook.com/intern/diff/D27691960)

[ghstack-poisoned]

zhxchen17 committed:
Update on "[JIT] Put explicit error message on class attribute accesses."


Summary: Resolving https://github.com/pytorch/pytorch/issues/51139

Test Plan: python test/test_jit.py TestClassType.test_unresolved_attributes

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27691960](https://our.internmc.facebook.com/intern/diff/D27691960)

[ghstack-poisoned]

gmagogsfm commented on torch/csrc/jit/frontend/tree_views.cpp:
```diff
@@ -0,0 +1,40 @@
+    for (const auto& expr : Assign{node.get()}.lhs_list()) {
+      collectUnresolvedNames(names, expr);
+    }
+  } else if (node.kind() == TK_TUPLE_LITERAL) {
+    for (const auto& expr : TupleLiteral{node.get()}.inputs()) {
+      collectUnresolvedNames(names, expr);
+    }
+  } else if (node.kind() == TK_LIST_LITERAL) {
+    for (const auto& expr : ListLiteral{node.get()}.inputs()) {
+      collectUnresolvedNames(names, expr);
+    }
```
Nice! Could you add test case covering these branches as well? including nested tuple literals and list literals? Thanks!

zhxchen17 commented on tools/build_variables.bzl:
```diff
@@ -155,6 +155,7 @@ core_sources_full_mobile = [
     "torch/csrc/jit/frontend/script_type_parser.cpp",
     "torch/csrc/jit/frontend/sugared_value.cpp",
     "torch/csrc/jit/frontend/tracer.cpp",
+    "torch/csrc/jit/frontend/tree_views.cpp",
```
to be honest I'm not sure why core_source_full_mobile contains many files from torch/csrc/jit/frontend, is there particular reason behind these?

zhxchen17 commented on test/test_jit.py:
```diff
@@ -4622,76 +4622,6 @@ def test(backward=False):
-        for tensor in tensors:
-            for i in range(len(script_funs)):
-                self.assertEqual(test_func(script_funs[i], x, tensor), test_func(funs[i], x, tensor))
-
```
How do you see this? I accidentally removed this code during rebase but fixed it in the next amend. Is this problematic?

zhxchen17 replied:
That's the case when I firstly saw this, but there's an actual conflict between my modification on this test and removal of this test from another commit yesterday, so I need remove this to resolve the conflict.

zhxchen17 committed:
Update on "[JIT] Put explicit error message on class attribute accesses."


Summary: Resolving https://github.com/pytorch/pytorch/issues/51139

Test Plan: python test/test_jit.py TestClassType.test_unresolved_attributes

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27691960](https://our.internmc.facebook.com/intern/diff/D27691960)

[ghstack-poisoned]

zhxchen17 commented:
AnnAssign, tweak bzl file, add more tests.

zhxchen17 committed:
Update on "[JIT] Put explicit error message on class attribute accesses."


Summary: Resolving https://github.com/pytorch/pytorch/issues/51139

Test Plan: python test/test_jit.py TestClassType.test_unresolved_attributes

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27691960](https://our.internmc.facebook.com/intern/diff/D27691960)

[ghstack-poisoned]

gmagogsfm commented on torch/jit/frontend.py:
```diff
@@ -157,6 +157,23 @@ def get_class_properties(cls, self_name):
+def get_class_assigns(ctx, cls_ast):
+    assigns = []
+
+    def build_assign(builder, entry):
```
```suggestion
    def maybe_build_assign(builder, entry):
```

gmagogsfm reviewed (approved):
Looking great!

zhxchen17 committed:
Update torch/jit/frontend.py

Co-authored-by: gmagogsfm <gmagogsfm@users.noreply.github.com>

zhxchen17 committed:
Update on "[JIT] Put explicit error message on class attribute accesses."


Summary: Resolving https://github.com/pytorch/pytorch/issues/51139

Test Plan: python test/test_jit.py TestClassType.test_unresolved_attributes

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27691960](https://our.internmc.facebook.com/intern/diff/D27691960)

[ghstack-poisoned]
